ibtk/include/ibtk/IBTK_MPI.h

```diff
@@ -0,0 +1,332 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef included_IBTK_MPI
+#define included_IBTK_MPI
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "SAMRAI_config.h"
+
+#include "mpi.h"
+
+#include "tbox/Utilities.h"
+#include <boost/iterator/iterator_concepts.hpp>
+
+namespace IBTK
+{
+inline MPI_Datatype
+mpi_type_id(const double)
+{
+    return MPI_DOUBLE;
+}
+
+inline MPI_Datatype
+mpi_type_id(const std::pair<double, int>)
+{
+    return MPI_DOUBLE_INT;
+}
+
+inline MPI_Datatype
+mpi_type_id(const int)
+{
+    return MPI_INT;
+}
+
+inline MPI_Datatype
+mpi_type_id(const std::pair<int, int>)
+{
+    return MPI_2INT;
+}
+
+inline MPI_Datatype
+mpi_type_id(const float)
+{
+    return MPI_FLOAT;
+}
+
+inline MPI_Datatype
+mpi_type_id(const std::pair<float, int>)
+{
+    return MPI_FLOAT_INT;
+}
+
+inline MPI_Datatype
+mpi_type_id(const char)
+{
+    return MPI_CHAR;
+}
+
+template <typename T>
+inline MPI_Datatype
+mpi_type_id(const T&)
+{
+    static_assert(!std::is_same<T, T>::value,
+                  "The given type does not have a corresponding MPI_Datatype value. At this time only char, int, "
+                  "unsigned int, float, double, std::pair<int, int>, std::pair<int, double>, and "
+                  "std::pair<int, float> are supported by IBTK_MPI.");
+    return MPI_CHAR;
+}
+
+/**
+ * @brief Provides C++ wrapper around MPI routines.
+ *
+ * The IBTK_MPI struct provides simple interfaces to common MPI routines. All function calls allow a communicator to be
+ * used. If a null communicator or none is present, the default communicator set by setCommunicator() is used.
+ *
+ * Note that this class is a utility class to group function calls in one
+ * name space (all calls are to static functions).  Thus, you should never
+ * attempt to instantiate a class of type IBTK_MPI; simply call the functions
+ * as static functions using the IBTK_MPI::function(...) syntax.
+ */
+
+struct IBTK_MPI
+{
+    /**
+     * MPI Types
+     */
+    using comm = MPI_Comm;
+    using group = MPI_Group;
+    using request = MPI_Request;
+    using status = MPI_Status;
+
+    /**
+     * Set the communicator that is used for the MPI communication routines.
+     * The default communicator is MPI_COMM_WORLD.
+     */
+    static void setCommunicator(IBTK_MPI::comm communicator);
+
+    /**
+     * Get the current MPI communicator.  The default communicator is
+     * MPI_COMM_WORLD.
+     */
+    static IBTK_MPI::comm getCommunicator();
+
+    /**
+     * Get SAMRAI World communicator.
+     */
+    static IBTK_MPI::comm getSAMRAIWorld();
+
+    /**
+     * Return the processor rank (identifier) from 0 through the number of
+     * processors minus one.
+     */
+    static int getRank(IBTK_MPI::comm communicator = MPI_COMM_NULL);
```

Comment:
I think we can get rid of `MPI_COMM_NULL` here and elsewhere by just writing `getCommunicator()` as the default value.